### ISSUE TITLE
system limit 32000000?

### ISSUE THREAD

**Issue by ggcking** [heavyai/heavydb#366]
I run sql with several join tables, it shows
"...would contain 191011399 rows, which is more than the current system limit of 32000000" , how to solve it ?

**Comment by cdessanti** [comment]
Being 32M the default size of a fragment, I guess a materialization is happening on a right table/subquery that needs more than 32M of rows to be materialized.
Just a guess, so I can be wrong. Which queries are you running?

**Comment by ggcking** [comment]
I run the query about several join and left join tables. And how can i do to exceed the 32M
 ?

**Comment by asuhan** [comment]
We need the table schema definitions and the queries in order to investigate this.

**Comment by ggcking** [comment]
I "create tables " as default.

**Comment by asuhan** [comment]
We can't investigate without knowing the column types and the queries you're trying to run.

**Comment by cdessanti** [comment]
I didn't say you created tables with a fragment size longer than 32m (you can and to get some performance boost you should also).
Depending on the query some intermediate materialization can occur; just a guess, and I can be wrong.
So the DDLs and the queries would help to dig the problem

**Comment by ggcking** [comment]
it is a long sql. 

select zb.billid, zb.ttime, mb.TRANSACTOR__dx, mb.LXDH, mb.jbbm__dx, 
zb.dwdh, 
dw.mc as dwdh_value, 
mb.BXSY, mb.BXZE, mb.ZFFS,
e.enumvalue as zffs_vale, 
mb.SQZFSJ, 
cb1.PAYMONEY, mb.HXJK, 
cb1.BAOXIAOREN__dx, 
dx3.dxmc as BAOXIAOREN_mc,
cb1.SSBM,dx4.dxmc as ssbm_mc, 
cb1.ywhd, dx5.dxmc as ywhd_mc,
cb1.fyxm, dx6.dxmc as fyxm_mc,
cb1.JTGJ, e2.enumvalue as jtgj_name,
cb1.SUBDEP, dx7.dxmc as SUBDEP_mc,  
cb1.PAYEE, dx8.dxmc as payee_mc,
cb1.SKFYHNW, dx9.dxmc as SKFYHNW_mc
from zhtable zb 
join maintable mb on mb.billid = zb.billid  and mb.dwdh = '021E' and mb.year_= 2019
join detailTable cb1 on cb1.ywid = mb.ywid 
left join t_xtdw dw on dw.dh = mb.dwdh
left join TENUM e on e.enumid = mb.zffs 
left join xt5113 dx1 on dx1.dxid = mb.TRANSACTOR__Dx 
left join xt5001 dx2 on dx2.dxid = mb.JBBM__dx 
left join xt5113 dx3 on dx3.dxid = cb1.BAOXIAOREN__dx 
left join xt5001 dx4 on dx4.dxid = cb1.SSBM__dx 
left join xt9622 dx5 on dx5.dxid = cb1.ywhd__dx 
left join xt9623 dx6 on dx6.dxid = cb1.fyxm__dx 
left join xt5001 dx7 on dx7.dxid = cb1.SUBDEP__dx 
left join xt5113 dx8 on dx8.dxid = cb1.PAYEE__dx 
left join xt5028 dx9 on dx9.dxid = cb1.SKFYHNW__dx 
where zb.zfbz = 0 and zb.year_ = 2019 and zb.dwdh = '021E' 
and e.enumlxid = '50002000'
order by zb.year_, zb.ttime;

--  zhtable  about 300,000 records
create table zhtable (
  billid bigint,
  year_ bigint,
  dwdh TEXT ENCODING DICT(32)
  ...
)
--  maintable  about 300,000 records
create table maintable (
	billid BIGINT,
	year_ INTEGER,
	dwdh TEXT ENCODING DICT(32),
	ywid INTEGER,
	TRANSACTOR__Dx INTEGER,
	JBBM__dx INTEGER
	...
)

--  detailTable about 3,000,000 records
create table detailTable(
	ywid INTEGER,
	BAOXIAOREN__dx INTEGER,
	SSBM__dx INTEGER,
	ywhd__dx INTEGER,
	fyxm__dx INTEGER,
	SUBDEP__dx INTEGER,
	PAYEE__dx INTEGER,
	SKFYHNW__dx INTEGER,
	...
)

--  TENUM about 3,000 records
create table TENUM(
	enumid TEXT ENCODING DICT(32),
	enumvalue TEXT ENCODING DICT(32)
)

--  t_xtdw about 1,000 records
create table t_xtdw(
	dh TEXT ENCODING DICT(32),
	mc TEXT ENCODING DICT(32)
	
)
--  xt[0..9] about 10,000 records
create table xt5113(
  dxid INTEGER,
  dxmc TEXT ENCODING DICT(32)
)

**Comment by asuhan** [comment]
@ggcking: Could you run a count query with the same joins and filters, that is:

```
select count(*)
from zhtable zb
join maintable mb on mb.billid = zb.billid and mb.dwdh = '021E' and mb.year_= 2019
join detailTable cb1 on cb1.ywid = mb.ywid
left join t_xtdw dw on dw.dh = mb.dwdh
left join TENUM e on e.enumid = mb.zffs
left join xt5113 dx1 on dx1.dxid = mb.TRANSACTOR__Dx
left join xt5001 dx2 on dx2.dxid = mb.JBBM__dx
left join xt5113 dx3 on dx3.dxid = cb1.BAOXIAOREN__dx
left join xt5001 dx4 on dx4.dxid = cb1.SSBM__dx
left join xt9622 dx5 on dx5.dxid = cb1.ywhd__dx
left join xt9623 dx6 on dx6.dxid = cb1.fyxm__dx
left join xt5001 dx7 on dx7.dxid = cb1.SUBDEP__dx
left join xt5113 dx8 on dx8.dxid = cb1.PAYEE__dx
left join xt5028 dx9 on dx9.dxid = cb1.SKFYHNW__dx
where zb.zfbz = 0 and zb.year_ = 2019 and zb.dwdh = '021E'
and e.enumlxid = '50002000';
```

? That way we'll know how many matches there are.

**Comment by ggcking** [comment]
it is about 3 millions records.

**Comment by asuhan** [comment]
Interesting, we should be able to complete the query then. Can you run `\version` in the omnisql console and tell us what it reports?

@alexbaden we might want to investigate this with synthetic data, looks like something which we should be able to fix. If the count query comes back at 3 million indeed, I worry that this might be related to the bump allocator, if this is a build after those changes.

**Comment by cdessanti** [comment]
If the main table has 3m of records and the result set returned is the same after multiple filtering it's likely intermediate materilization with more than 3m are happening.

Ottieni Outlook per Android<https://aka.ms/ghei36>

________________________________

**Comment by asuhan** [comment]
@cdessanti There is nothing in the query which would led me to believe that we would materialize anything except for the end result, so I'm not really sure what's going on.

**Comment by cdessanti** [comment]
Maybe here.

from zhtable zb join maintable mb on mb.billid = zb.billid and mb.dwdh = '021E' and mb.year_= 2019 (the same filter on the left table is happening also in the where clause)

If the bill is unique, no problem, but if not the filtering is pushed at the end of the plan execution, so a many to many join is happening.

Ottieni Outlook per Android<https://aka.ms/ghei36>

________________________________

**Comment by asuhan** [comment]
No, we're not supposed to materialize anything except for the end result, even if the join isn't one-to-one. If we do, it's something we could and should fix.

**Comment by ggcking** [comment]
@asuhan the \version
OmniSci Server Version: 4.7.0-20190614-4559864637
@cdessanti the billid is unique

**Comment by ggcking** [comment]
bwt: the omnisci server run on docker

**Comment by ggcking** [comment]
btw: the omnisci server run on docker

**Comment by baretomas** [comment]
I got this as well on a query where I join a table with itself. It's supposed to generate a lot of data, but I have to worry about intermediate materialization?  How am I supposed to avoid that on a growing database?

Is it even realistic to do self joins on big tables with omnisci?

**Comment by dipanwita2019** [comment]
i faced the same issue ...any solution or answer to this ?

**Comment by psilvest** [comment]
This issue has been inactive for a while, and the project has changed substantially. If this still occurs on the latest release, please add reproduction steps and environment details.
